File: Graph/php/longestPath.php
<?php

declare(strict_types=1);

function longestPath(array $graph): int
{
    $distance = [];
    foreach ($graph as $node => $neighbors) {
        if (!$neighbors) {
            $distance[$node] = 0;
        }
    }

    $maxLength = 0;
    foreach ($graph as $node => $_) {
        $value = traverseDistance($graph, (string) $node, $distance);
        if ($value > $maxLength) {
            $maxLength = $value;
        }
    }

    return $maxLength;
}

function traverseDistance(array $graph, string $node, array &$distance): int
{
    if (isset($distance[$node])) {
        return $distance[$node];
    }

    $maxNeighbor = 0;
    foreach ($graph[$node] ?? [] as $neighbor) {
        $candidate = traverseDistance($graph, (string) $neighbor, $distance);
        if ($candidate > $maxNeighbor) {
            $maxNeighbor = $candidate;
        }
    }

    $distance[$node] = $maxNeighbor + 1;
    return $distance[$node];
}

File: Graph/csharp/LongestPath.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class LongestPath
{
    public static int Solve(Dictionary<string, List<string>> graph)
    {
        var distance = new Dictionary<string, int>();
        foreach (var (node, neighbors) in graph)
        {
            if (neighbors.Count == 0)
            {
                distance[node] = 0;
            }
        }

        var maxLength = 0;
        foreach (var node in graph.Keys)
        {
            var value = Traverse(graph, node, distance);
            if (value > maxLength)
            {
                maxLength = value;
            }
        }

        return maxLength;
    }

    private static int Traverse(Dictionary<string, List<string>> graph, string node, Dictionary<string, int> distance)
    {
        if (distance.TryGetValue(node, out var cached))
        {
            return cached;
        }

        var maxNeighbor = 0;
        if (graph.TryGetValue(node, out var neighbors))
        {
            foreach (var neighbor in neighbors)
            {
                var candidate = Traverse(graph, neighbor, distance);
                if (candidate > maxNeighbor)
                {
                    maxNeighbor = candidate;
                }
            }
        }

        distance[node] = maxNeighbor + 1;
        return distance[node];
    }
}
